Stable names for lambda lifted method and fresh names

retronym opened the pull request:
Fresh names are created using a FreshNameCreator, which appends
an increasing number to the given prefix.

```
scala> val fresh = new scala.reflect.internal.util.FreshNameCreator()
fresh: scala.reflect.internal.util.FreshNameCreator = scala.reflect.internal.util.FreshNameCreator@42b84286

scala> List("foo$", "bar$", "foo$").map(fresh.newName(_))
res1: List[String] = List(foo$1, bar$1, foo$2)
```

Each compilation unit had its own fresh name creator, which is used
in the regular compiler. Macros and quasiquotes make use of a global
creator (at least, as of #3401).

Both of these are too broadly scoped to help achieve deterministic
fresh names: if sources are recompiled in a different order or separately
recompiled, the fresh name counters can be different. Methods in a given
compilation unit are not necessarily typechecked in a linear fashion;
they might be typechecked ahead of time to provide an inferred type
to a caller.

This commit:

  - Changes all known fresh name creations within the typer phase (in which
    out-of-order typechecking is a factor) to use a fineer grained fresh
    name creator. How fine grained? A fresh name generated as some position
    `p` shares the fresh name generator scoped at the closest method or
    class that encloses that the outermost enclosing tree at the same
    position. This definition is designed to give a shared fresh name
    creator for all fresh names generated in `macro1(macro2())`, even if
    the fresh names are requiested from with a Typer in the macro enclosed
    by a synthetic method.
  - Changes macro fresh names to use the same fresh naming scheme as the regular
    typechecker. An opt-out compiler option allows the old behaviour, but I'm
    interested to find real-world cases where the new scheme actually causes
    a problem

In addition, a small change is made to lambda lift to lift local methods in the
order that they are encountered during traversal, rather than sorting them
based on `Symbol.isLess` (which include `Symbol.id`, an order-of-typechecking
dependent value).

References scala/scala-dev#405

retronym committed:
Stable names for lambda lifted method and fresh names

Fresh names are created using a FreshNameCreator, which appends
an increasing number to the given prefix.

```
scala> val fresh = new scala.reflect.internal.util.FreshNameCreator()
fresh: scala.reflect.internal.util.FreshNameCreator = scala.reflect.internal.util.FreshNameCreator@42b84286

scala> List("foo$", "bar$", "foo$").map(fresh.newName(_))
res1: List[String] = List(foo$1, bar$1, foo$2)
```

Each compilation unit had its own fresh name creator, which is used
in the regular compiler. Macros and quasiquotes make use of a global
creator (at least, as of #3401).

Both of these are too broadly scoped to help achieve deterministic
fresh names: if sources are recompiled in a different order or separately
recompiled, the fresh name counters can be different. Methods in a given
compilation unit are not necessarily typechecked in a linear fashion;
they might be typechecked ahead of time to provide an inferred type
to a caller.

This commit:

  - Changes all known fresh name creations within the typer phase (in which
    out-of-order typechecking is a factor) to use a fineer grained fresh
    name creator. How fine grained? A fresh name generated as some position
    `p` shares the fresh name generator scoped at the closest method or
    class that encloses that the outermost enclosing tree at the same
    position. This definition is designed to give a shared fresh name
    creator for all fresh names generated in `macro1(macro2())`, even if
    the fresh names are requiested from with a Typer in the macro enclosed
    by a synthetic method.
  - Changes macro fresh names to use the same fresh naming scheme as the regular
    typechecker. An opt-out compiler option allows the old behaviour, but I'm
    interested to find real-world cases where the new scheme actually causes
    a problem

In addition, a small change is made to lambda lift to lift local methods in the
order that they are encountered during traversal, rather than sorting them
based on `Symbol.isLess` (which include `Symbol.id`, an order-of-typechecking
dependent value).

retronym committed:
Report an error on problematic use of unit.fresh

I've used this to flush out the corner cases fixed in the previous commit.

retronym committed:
Tone error down to a dev warning

retronym commented:
/synch

retronym commented:
/rebuild

retronym commented:
This causes a binary incompatiblity with partest which is build with the full Scala optimizer enabled.

```
[test] [error] Uncaught exception when running partest: java.lang.NoSuchMethodError: scala.reflect.internal.util.Collections.$anonfun$distinctBy$1(Lscala/Function1;Lscala/collection/mutable/ListBuffer;Lscala/collection/mutable/Set;Ljava/lang/Object;)Ljava/lang/Object;
[test] sbt.ForkMain$ForkError: java.lang.NoSuchMethodError: scala.reflect.internal.util.Collections.$anonfun$distinctBy$1(Lscala/Function1;Lscala/collection/mutable/ListBuffer;Lscala/collection/mutable/Set;Ljava/lang/Object;)Ljava/lang/Object;
[test] 	at scala.tools.partest.nest.AbstractRunner.run(AbstractRunner.scala:160)
[test] 	at scala.tools.partest.sbt.PartestTask.execute(Framework.scala:43)
[test] 	at sbt.ForkMain$Run$2.call(ForkMain.java:296)
[test] 	at sbt.ForkMain$Run$2.call(ForkMain.java:286)
[test] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[test] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[test] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[test] 	at java.lang.Thread.run(Thread.java:748)
```

I'll release a new version of partest that only enables local optimizations.

retronym commented:
Rebased on top of a new version of partest.

retronym commented:
Rebased on the yet-to-be-merged #6391 to speed up the rebuilds this needs.

retronym commented:
The test failures suggest that the test suite under `./scalacheck` is being compiled against, and inlining from, `./library`, but being executed with STARR on the classpath.

```
java.lang.BootstrapMethodError: java.lang.NoSuchMethodError: scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(Lscala/runtime/ObjectRef;Lscala/Function2;Ljava/lang/Object;)Ljava/lang/Object;
	at TreeMapTest$.$anonfun$new$110(treemap.scala:151)
	at TreeMapTest$.$anonfun$new$110$adapted(treemap.scala:150)
	at scala.Function1.$anonfun$andThen$1(Function1.scala:52)
```

```
> show scalacheck/autoScalaLibrary
[info] false
> show scalacheck/test:autoScalaLibrary
[info] false
> show scalacheck/test:full-classpath
[info] * Attributed(/Users/jz/code/scala/target/scalacheck/test-classes)
[info] * Attributed(/Users/jz/code/scala/build/quick/classes/scalacheck)
[info] * Attributed(/Users/jz/code/scala/build/quick/classes/library)
[info] * Attributed(/Users/jz/code/scala/build/quick/classes/reflect)
[info] * Attributed(/Users/jz/code/scala/build/quick/classes/compiler)
[info] * Attributed(/Users/jz/code/scala/build/quick/classes/scaladoc)
[info] * Attributed(/Users/jz/.ivy2/cache/org.apache.ant/ant/jars/ant-1.9.4.jar)
[info] * Attributed(/Users/jz/.ivy2/cache/org.apache.ant/ant-launcher/jars/ant-launcher-1.9.4.jar)
[info] * Attributed(/Users/jz/.ivy2/cache/org.scala-lang.modules/scala-asm/bundles/scala-asm-6.0.0-scala-1.jar)
[info] * Attributed(/Users/jz/.ivy2/cache/org.scala-lang.modules/scala-xml_2.13.0-M3/bundles/scala-xml_2.13.0-M3-1.1.0.jar)
[info] * Attributed(/Users/jz/.ivy2/cache/jline/jline/jars/jline-2.14.5.jar)
[info] * Attributed(/Users/jz/.ivy2/cache/org.scalacheck/scalacheck_2.13.0-M1/jars/scalacheck_2.13.0-M1-1.13.5.jar)
[info] * Attributed(/Users/jz/.ivy2/cache/org.scala-sbt/test-interface/jars/test-interface-1.0.jar)
[success] Total time: 0 s, completed 13/03/2018 9:20:12 AM
> show scalacheck/test:console
[info] Starting scala interpreter...
[info]
Welcome to Scala 2.13.0-M3 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_162).
Type in expressions for evaluation. Or try :help.

scala> classOf[scala.collection.TraversableOnce[_]].getProtectionDomain.getCodeSource
res0: java.security.CodeSource = (file:/Users/jz/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.13.0-M3.jar <no signer certificates>)
```

retronym commented:
Forking fixes the classpath issue, but no longer reports failures:

```
> scalacheck/testOnly TreeMapTest
[info] Compiling 1 Scala source to /Users/jz/code/scala/target/scalacheck/test-classes...
[info] + TreeMap.splitAt: OK, passed 100 tests.
[info] + TreeMap.init/last identity: OK, passed 100 tests.
[info] ! TreeMap.faily: Falsified after 0 passed tests.
[info] > ARG_0: 0
[info] > ARG_0_ORIGINAL: -1
[info] + TreeMap.span identity: OK, passed 100 tests.
[info] + TreeMap.remove all: OK, passed 100 tests.
[info] + TreeMap.contains all: OK, passed 100 tests.
[info] + TreeMap.takeWhile: OK, passed 100 tests.
[info] + TreeMap.until is exclusive: OK, passed 100 tests.
[info] + TreeMap.dropWhile: OK, passed 100 tests.
[info] + TreeMap.drop: OK, passed 100 tests.
[info] + TreeMap.take: OK, passed 100 tests.
[info] + TreeMap.worst-case tree height is iterable: OK, passed 100 tests.
[info] + TreeMap.foreach/iterator consistency: OK, passed 100 tests.
[info] + TreeMap.sorted: OK, passed 100 tests.
[info] + TreeMap.from is inclusive: OK, passed 100 tests.
[info] + TreeMap.to is inclusive: OK, passed 100 tests.
[info] + TreeMap.head: OK, passed 100 tests.
[info] + TreeMap.remove single: OK, passed 100 tests.
[info] + TreeMap.slice: OK, passed 100 tests.
[info] + TreeMap.size: OK, passed 100 tests.
[info] + TreeMap.last: OK, passed 100 tests.
[info] + TreeMap.head/tail identity: OK, passed 100 tests.
[info] + TreeMap.toSeq: OK, passed 100 tests.
[info] + TreeMap.take/drop identity: OK, passed 100 tests.
[info] Passed: Total 1, Failed 0, Errors 0, Passed 1
[success] Total time: 9 s, completed 13/03/2018 9:33:58 AM
```

71f9b43c9eb9a3bb8ee16619c432da35609f23bc

retronym commented:
[Reported](https://github.com/sbt/zinc/pull/505/files#r173996657) our desire to leave STARR off the non-forked test classpath to @eed3si9n who is currently refactoring that part of SBT.

eed3si9n commented:
I opened https://github.com/sbt/sbt/issues/4009. Hopefully this captures what's going on.

retronym commented:
Cherry-picked #6440 (submitted to 2.12.x) into this PR to unblock.

retronym commented:
The tripwire I set in 00d1a5ae441cbbc3c7c3541092032586bf8ab576 has triggered for some tests (mostly around toolbox compilation, I think). So there are a few more problematic uses of fresh names that need attention.

retronym commented:
Rebased

retronym commented:
/rebuild

adriaanm commented:
Looks like there's still a conflict

retronym commented:
In addition to the new merge conflict, there is also a semantatic conflict.

Looks like the recent changes to collections serialization are somewhat fragile with the way that factories are serialized. We shouldn't rely on anonymous class names in things that are supposed to be serialization compatible:

```
sbt.ForkMain$ForkError: java.lang.ClassNotFoundException: scala.collection.IterableFactory$$anon$1
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at java.io.ObjectInputStream.resolveClass(ObjectInputStream.java:686)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1866)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1749)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2040)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1571)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2285)
	at java.io.ObjectInputStream.defaultReadObject(ObjectInputStream.java:561)
	at scala.collection.generic.DefaultSerializationProxy.readObject(DefaultSerializationProxy.scala:35)
	at sun.reflect.GeneratedMethodAccessor3.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1158)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2176)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2067)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1571)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
	at scala.SerializationStability$.deserialize(SerializationStabilityTest.scala:62)
	at scala.SerializationStability$.main(SerializationStabilityTest.scala:79)
	at scala.SerializationStabilityTest.testAll(SerializationStabilityTest.scala:221)
```

retronym commented:
@szeiger This will certainly be a lot better after this PR, so we could just reset the test. But I wonder how many factories we actually have that are anonymous?

The ones this test failure found are among:

```scala
  implicit def toFactory[A, CC[_]](factory: IterableFactory[CC]): Factory[A, CC[A]] =
    new Factory[A, CC[A]] with Serializable {
      def fromSpecific(it: IterableOnce[A]): CC[A] = factory.from[A](it)
      def newBuilder: Builder[A, CC[A]] = factory.newBuilder[A]
    }

  implicit def toBuildFrom[A, CC[_]](factory: IterableFactory[CC]): BuildFrom[Any, A, CC[A]] =
    new BuildFrom[Any, A, CC[A]] {
      def fromSpecificIterable(from: Any)(it: Iterable[A]) = factory.from(it)
      def newBuilder(from: Any) = factory.newBuilder
    }
```

retronym commented:
A search for `new Factory.*\{` has 15 matches.

retronym committed:
Use named classes or objects for collection BuildFroms

For consistency with the treatment of Factory.

I also spotted some missing `extends Serializable` on string
and array factories and added them.

retronym commented:
@szeiger I've preceded this PR with a commit to switch to named classes for `Factory`-s and `BuildFrom`-s. I took the opportunity to use reusable instances where possible.

retronym commented:
Ah, I see basically the same thing has just happened on 2.13.x

szeiger commented:
Yes, I already did the changes for `Factory`. But it's easy to accidentally (or simply to avoid the overhead) serialize closures. We still rely on this for views and for lazy lists.

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -754,7 +754,12 @@ trait Macros extends MacroRuntimes with Traces with Helpers {
+  def macroExpand(typer: Typer, expandee: Tree, mode: Mode, pt: Type): Tree = {
+    // By default, use the current typer's fresh name creator in macros. The compiler option
+    // allows people to opt in to the old behaviour of Scala 2.12, which used a global fresh creator.
+    if (!settings.YmacroFresh.value) currentFreshNameCreator = typer.fresh
```
could we use -Xsource:2.12 instead?

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Contexts.scala:
```diff
@@ -637,6 +637,18 @@ trait Contexts { self: Analyzer =>
     def nextEnclosing(p: Context => Boolean): Context =
       if (p(this)) this else outer.nextEnclosing(p)
 
+    final def outermostContextAtCurrentPos: Context = {
```
needs a comment, can be taken from PR description (what does it compute, and why). Avoid having this drift, co-opted for other incompatible purposes.

adriaanm commented on test/junit/scala/tools/nsc/DeterminismTest.scala:
```diff
@@ -0,0 +1,228 @@
+package scala.tools.nsc
```
💯 awesome way to test stability, but it could really use a comment

retronym committed:
Merge remote-tracking branch 'origin/2.13.x' into topic/stable-fresh

 Conflicts:
	test/junit/scala/SerializationStabilityTest.scala

retronym commented:
The build failure is just the macro annotation binary incompat on Travis for the second-to-last commit. It can be safely ignored.

adriaanm reviewed (commented):
(Hmm, looks like I got interrupted before submitting my review :-/)

retronym commented:
I don’t think Xsource make sense here. The option is to give an escape
hatch in case of a bug. But this isn’t a change to the language (like
overloading or inference), but a change to a detail of code generation that
unfortunately might be observable when we have change maybe compiler
plugins and macros around.

retronym commented:
*this option
On Thu, 7 Jun 2018 at 21:28, Jason Zaugg <jzaugg@gmail.com> wrote:

> I don’t think Xsource make sense here. The option is to give an escape
> hatch in case of a bug. But this isn’t a change to the language (like
> overloading or inference), but a change to a detail of code generation that
> unfortunately might be observable when we have change maybe compiler
> plugins and macros around.
>

raboof commented:
This looks great!

We compiled Akka 2.5.22 with Scala 2.13.0-M5 (which contains this improvement). Here I see a difference between our published artifact and my (independent) local build that suggests there might be more sources of nondeterminism in this area.

The diffoscope report can be found at https://arnout.engelen.eu/rb/akka/2.13.0-M5/2.5.22/reproducible-builds-diffoscope-output-akka-actor_2.13.0-M5-2.5.22.jar/index.html - for example see `akka/io/InetAddressDnsResolver.class`

raboof commented:
> I see a difference between our published artifact and my (independent) local build

I tracked the change down and it was due to a compiler plugin, scalac seems fine! Sorry for the noise and great work here :smile:

adriaanm commented:
Great! Thanks for investigating.
